### PR TITLE
Support auto_fp16 using torch.cuda.amp

### PR DESCRIPTION
**Pull request by ycxioooong**
Testing using mmdet faster_rcnn_r50_fpn_fp16_1x_coco.

Ubuntu 18.04
torch1.7.1
CUDA10.2
GPU: 2080Ti

Train speed 0.3s/iter;
Memory: 2.8GB;
Inference: 99fps;
mAP 37.4 (compared with previous fp16 implementation 37.5)

### CODE REVIEW

File: mmcv/runner/hooks/optimizer.py [hunk 1]
```diff
@@ -125,37 +167,50 @@ def after_train_iter(self, runner):
...
+
+        if self._use_torch_amp:
+            self.loss_scaler.scale(runner.outputs['loss']).backward()
+            self.loss_scaler.unscale_(runner.optimizer)
+            # grad clip
             if self.grad_clip is not None:
-                grad_norm = self.clip_grads(fp32_weights)
-                if grad_norm is not None:
-                    # Add grad norm to the logger
-                    runner.log_buffer.update({'grad_norm': float(grad_norm)},
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/optimizer.py:152]: We may keep the logging of gradient norm

File: mmcv/runner/hooks/optimizer.py [hunk 2]
```diff
@@ -66,37 +89,56 @@ def __init__(self,
...
-            self.loss_scaler = LossScaler(init_scale=loss_scale, mode='static')
+            self._scale_update_param = loss_scale
+            if self._use_torch_amp:
+                self.loss_scaler = GradScaler(init_scale=loss_scale)
+            else:
+                self.loss_scaler = LossScaler(
+                    init_scale=loss_scale, mode='static')
         elif isinstance(loss_scale, dict):
-            self.loss_scaler = LossScaler(**loss_scale)
+            self._scale_update_param = None
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/optimizer.py:109]: Can we simplify the relationship between `self._scale_update_param` and `self.loss_scaler`?

Comment (ycxioooong) [reply mmcv/runner/hooks/optimizer.py:109]: Good idea. I have another option in mind, will have a try.

**Comment by hhaAndroid** [comment]
@ycxioooong @ZwwWayne  `dy` is mean `dynamic`

===================================
configs/fp16/mask_rcnn_r50_fpn_fp16_1x_coco.py    {'memory': 3023, 'bbox_mAP': 0.38, 'segm_mAP': 0.345}
configs/fp16/mask_rcnn_r50_fpn_fp16_1x_coco_dy.py    {'memory': 3019, 'bbox_mAP': 0.381, 'segm_mAP': 0.347}

configs/fp16/faster_rcnn_r50_fpn_fp16_1x_coco.py    {'memory': 2776, 'bbox_mAP': 0.372}
configs/fp16/faster_rcnn_r50_fpn_fp16_1x_coco_dy.py    {'memory': 2775, 'bbox_mAP': 0.375}

configs/fp16/retinanet_r50_fpn_fp16_1x_coco.py    {'memory': 2684, 'bbox_mAP': 0.364}
configs/fp16/retinanet_r50_fpn_fp16_1x_coco_dy.py   {'memory': 2684, 'bbox_mAP': 0.367}

===================================

**Comment by ZwwWayne** [comment]
> @ycxioooong @ZwwWayne `dy` is mean `dynamic`
> 
> ===================================
> configs/fp16/mask_rcnn_r50_fpn_fp16_1x_coco.py {'memory': 3023, 'bbox_mAP': 0.38, 'segm_mAP': 0.345}
> configs/fp16/mask_rcnn_r50_fpn_fp16_1x_coco_dy.py {'memory': 3019, 'bbox_mAP': 0.381, 'segm_mAP': 0.347}
> 
> configs/fp16/faster_rcnn_r50_fpn_fp16_1x_coco.py {'memory': 2776, 'bbox_mAP': 0.372}
> configs/fp16/faster_rcnn_r50_fpn_fp16_1x_coco_dy.py {'memory': 2775, 'bbox_mAP': 0.375}
> 
> configs/fp16/retinanet_r50_fpn_fp16_1x_coco.py {'memory': 2684, 'bbox_mAP': 0.364}
> configs/fp16/retinanet_r50_fpn_fp16_1x_coco_dy.py {'memory': 2684, 'bbox_mAP': 0.367}
> 
> ===================================

LGTM.

**Review by ZwwWayne — APPROVED** [review]

**Comment by hhaAndroid** [comment]
GPU: V100 

===================================
configs/fp16/mask_rcnn_r50_fpn_fp16_1x_coco.py {'memory': 3023, 'bbox_mAP': 0.38, 'segm_mAP': 0.345, average iter time: 0.2876 s/iter}
configs/fp16/mask_rcnn_r50_fpn_fp16_1x_coco_dy.py {'memory': 3019, 'bbox_mAP': 0.381, 'segm_mAP': 0.347, average iter time: 0.3025 s/iter}

configs/fp16/faster_rcnn_r50_fpn_fp16_1x_coco.py {'memory': 2776, 'bbox_mAP': 0.372, average iter time: 0.1948 s/iter}
configs/fp16/faster_rcnn_r50_fpn_fp16_1x_coco_dy.py {'memory': 2775, 'bbox_mAP': 0.375, average iter time: 0.1954 s/iter}

configs/fp16/retinanet_r50_fpn_fp16_1x_coco.py {'memory': 2684, 'bbox_mAP': 0.364,average iter time: 0.1733 s/iter
}
configs/fp16/retinanet_r50_fpn_fp16_1x_coco_dy.py {'memory': 2684, 'bbox_mAP': 0.367,average iter time: 0.1730 s/iter
}
===================================

File: mmcv/runner/fp16_utils.py [hunk 3]
```diff
@@ -7,8 +7,14 @@
 import torch
 import torch.nn as nn
 
+from mmcv.utils import TORCH_VERSION
 from .dist_utils import allreduce_grads as _allreduce_grads
 
+try:
+    from torch.cuda.amp import autocast
+except ImportError:
+    pass
```
Comment (hellock) [inline mmcv/runner/fp16_utils.py:16]: Add a comment here to indicate the version requirement

File: mmcv/runner/hooks/optimizer.py [hunk 4]
```diff
@@ -47,13 +56,27 @@ class Fp16OptimizerHook(OptimizerHook):
...
         loss_scale (float | str | dict): Scale factor multiplied with loss.
             If loss_scale is a float, static loss scaling will be used with
             the specified scale. If loss_scale is a string, it must be
             'dynamic', then dynamic loss scaling will be used.
             It can also be a dict containing arguments of LossScaler.
-            Defaults to 512.
+            Defaults to 512. For Pytorch >= 1.6, mmcv uses official
+            implementation of GradScaler. If you use a dict version of
+            loss_scale to create GradScaler, plese refer to:
+
```
Comment (hellock) [inline mmcv/runner/hooks/optimizer.py:70]: These blank lines may break the docstring style.

File: mmcv/runner/hooks/optimizer.py [hunk 5]
```diff
@@ -38,7 +44,10 @@ def after_train_iter(self, runner):
 class Fp16OptimizerHook(OptimizerHook):
```
Comment (hellock) [inline mmcv/runner/hooks/optimizer.py:44]: Since the logic of torch.amp and our own implementation is quite different, we may implement two classes.
```python
if TORCH_VERSION != 'parrots' and TORCH_VERSION >= '1.6.0':
    class Fp16OptimizerHook:
        # amp implementation
else:
    class Fp16OptimizerHook:
        # our own implementation
```
It should be clearer for users.

**Comment by ZwwWayne** [comment]
Kindly ping @ycxioooong
